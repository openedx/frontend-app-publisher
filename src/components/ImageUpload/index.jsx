import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import StatusAlert from '../StatusAlert';

class ImageUpload extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.input.value,
      sizeValidationError: '',
    };

    this.handleFilePicked = this.handleFilePicked.bind(this);
    this.sizeValidator = this.sizeValidator.bind(this);
  }

  updateValue(value) {
    this.setState({ value });
    this.props.input.onChange(value);
  }

  handleFilePicked(event) {
    const {
      maxImageSizeKilo,
      requiredWidth,
      requiredHeight,
    } = this.props;
    const reader = new FileReader();
    const file = event.target.files[0];

    // file size is in bytes so we need to convert for our validation check
    if (file && file.size > maxImageSizeKilo * 1000) {
      this.updateValue('');
      this.setState({
        sizeValidationError: `That image is too large. Please upload an image that
        is less than ${maxImageSizeKilo} kB. Remember that image dimensions must be
        exactly ${requiredWidth}×${requiredHeight} pixels.`,
      });
      return;
    }

    reader.onload = () => {
      this.updateValue(reader.result);
    };
    if (file) {
      reader.readAsDataURL(file);
    }
  }

  sizeValidator(event) {
    const {
      requiredWidth,
      requiredHeight,
      meta: {
        pristine,
      },
    } = this.props;

    if (!pristine && (event.target.naturalWidth !== requiredWidth
        || event.target.naturalHeight !== requiredHeight)) {
      this.updateValue('');
      this.setState({
        sizeValidationError: `That image has the wrong dimensions. Please upload
        an image with exactly ${requiredWidth}×${requiredHeight} pixels.`,
      });
    } else {
      this.setState({ sizeValidationError: '' });
    }
  }

  render() {
    const {
      className,
      disabled,
      id,
      input: {
        name,
      },
      label,
      meta: {
        error,
        submitFailed,
      },
    } = this.props;

    const { sizeValidationError } = this.state;

    return (
      <div className={classNames('form-group', className)}>
        <div name={name} tabIndex={-1}>
          <label htmlFor={id} className="w-100 p-0">{label}</label>  {/* eslint-disable-line jsx-a11y/label-has-for */}
          {sizeValidationError
            && (
            <StatusAlert
              alertType="warning"
              message={sizeValidationError}
            />
            )}
          {submitFailed && error
            && (
            <StatusAlert
              alertType="danger"
              message={error}
            />
            )}
          <img src={this.state.value} alt="" className="uploaded-image" onLoad={this.sizeValidator} />
          <input
            id={id}
            type="file"
            accept="image/jpeg, image/png"
            onChange={this.handleFilePicked}
            disabled={disabled}
          />
        </div>
      </div>
    );
  }
}

ImageUpload.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string.isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  maxImageSizeKilo: PropTypes.number.isRequired,
  requiredWidth: PropTypes.number.isRequired,
  requiredHeight: PropTypes.number.isRequired,
  input: PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.string,
    onChange: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    error: PropTypes.string,
    pristine: PropTypes.bool,
    submitFailed: PropTypes.bool,
  }).isRequired,
};

ImageUpload.defaultProps = {
  className: '',
  disabled: false,
};

export default ImageUpload;
