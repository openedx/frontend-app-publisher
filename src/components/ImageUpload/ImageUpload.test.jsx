import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import ImageUpload from './index';

describe('ImageUpload', () => {
  it('shows an image upload with no prior image', () => {
    const component = shallow(<ImageUpload
      label="Course Image: *"
      id="image-no-prior"
      input={{
        name: 'imageSrc',
        value: null,
        onChange: () => null,
      }}
      meta={{
        error: null,
        pristine: true,
        submitFailed: false,
      }}
      maxImageSize={1000000}
      requiredWidth={1134}
      requiredHeight={675}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows an image upload with prior image', () => {
    const component = shallow(<ImageUpload
      label={<strong>Course Image: *</strong>}
      id="image-with-prior"
      input={{
        name: 'imageSrc',
        value: 'http://www.image.uploaded.source/view/image',
        onChange: () => null,
      }}
      meta={{
        error: null,
        pristine: true,
        submitFailed: false,
      }}
      maxImageSize={256000}
      requiredHeight={110}
      requiredWidth={110}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows an image upload and an error', () => {
    const component = shallow(<ImageUpload
      label="Image Error Test"
      id="image-with-error"
      input={{
        name: 'imageSrc',
        value: null,
        onChange: () => null,
      }}
      meta={{
        error: 'Required',
        pristine: false,
        submitFailed: true,
      }}
      maxImageSize={1000000}
      requiredWidth={1134}
      requiredHeight={675}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});
