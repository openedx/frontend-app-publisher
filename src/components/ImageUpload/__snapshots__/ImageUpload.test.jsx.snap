// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUpload shows an image upload and an error 1`] = `
<div
  className="form-group"
>
  <div
    name="imageSrc"
    tabIndex={-1}
  >
    <label
      className="w-100 p-0"
      htmlFor="image-with-error"
    >
      Image Error Test
    </label>
      
    <StatusAlert
      alertType="danger"
      className=""
      dismissible={false}
      iconClassNames={Array []}
      message="Required"
      onClose={[Function]}
      title={null}
    />
    <img
      alt=""
      className="uploaded-image"
      onLoad={[Function]}
      src={null}
    />
    <input
      accept="image/jpeg, image/png"
      disabled={false}
      id="image-with-error"
      onChange={[Function]}
      type="file"
    />
  </div>
</div>
`;

exports[`ImageUpload shows an image upload with no prior image 1`] = `
<div
  className="form-group"
>
  <div
    name="imageSrc"
    tabIndex={-1}
  >
    <label
      className="w-100 p-0"
      htmlFor="image-no-prior"
    >
      Course Image: *
    </label>
      
    <img
      alt=""
      className="uploaded-image"
      onLoad={[Function]}
      src={null}
    />
    <input
      accept="image/jpeg, image/png"
      disabled={false}
      id="image-no-prior"
      onChange={[Function]}
      type="file"
    />
  </div>
</div>
`;

exports[`ImageUpload shows an image upload with prior image 1`] = `
<div
  className="form-group"
>
  <div
    name="imageSrc"
    tabIndex={-1}
  >
    <label
      className="w-100 p-0"
      htmlFor="image-with-prior"
    >
      <strong>
        Course Image: *
      </strong>
    </label>
      
    <img
      alt=""
      className="uploaded-image"
      onLoad={[Function]}
      src="http://www.image.uploaded.source/view/image"
    />
    <input
      accept="image/jpeg, image/png"
      disabled={false}
      id="image-with-prior"
      onChange={[Function]}
      type="file"
    />
  </div>
</div>
`;
